### ISSUE TITLE
Client does not retransmit messages or time out

### ISSUE THREAD

**Issue by ksrm** [Tanganelli/CoAPthon#27]
I am using the HelperClient class. It seems to only send a single datagram and then wait forever for a response.

**Comment by Tanganelli** [comment]
Hi,
can you provide a detailed description of your environment? Which kind of request are you trying to achieve (GET/POST/PUT/DELETE)? how do you use the helper client?

**Comment by b-schuetz** [comment]
Hello,

I have the same issue when using blockwise transfer on a lossy link.

Maybe this helps to recreate the problem:
Here is the terminal output:

Server:
_python coapserver.py
CoAP Server start on 127.0.0.1:5683
['/basic', '/storage', '/child', '/separate', '/etag', '/', '/big', '/encoding', '/void', '/long', '/xml']_

Client:	
_python coapclient.py -o GET -p coap://127.0.0.1:5683/big
2016-04-22 10:52:40,767 - Thread-1   - coapthon.client.coap - DEBUG - Start receiver Thread
2016-04-22 10:52:40,775 - MainThread - coapthon.layers.messagelayer - DEBUG - send_request - From None, To ('127.0.0.1', 5683), None-None, GET-None, [Uri-Path: big, ] No payload
2016-04-22 10:52:40,775 - MainThread - coapthon.client.coap - DEBUG - send_datagram - From None, To ('127.0.0.1', 5683), CON-50247, GET-None, [Uri-Path: big, ] No payload_

The application is running on ubuntu.
I used netem to create a lossy enviroment by:
_sudo tc qdisc change dev lo root netem loss 10% delay 100ms_	

Sometimes the data is succesfull transferred, sometimes not. This indicates problems with the retransmission of lost messages.


Thank you for your great application, i really appreciate your good work.

Greetings
Bertram

**Comment by Tanganelli** [comment]
Interesting I will try to reproduce the issue this afternoon. Thanks in advance.

**Comment by lbourdain** [comment]
I think, you can modifiy the file helperclient.py and the definition of get() with this code 
```python
    def get(self, path, callback=None):  # pragma: no cover
        request = Request()
        request.destination = self.server
        request.code = defines.Codes.GET.number
        request.uri_path = path

        if callback is not None:
            thread = threading.Thread(target=self._thread_body, args=(request, callback))
            thread.start()
        else:
            delta = 1
            test = 0
            while test == 0:
                try:
                    self.protocol.send_message(request)
                    response = self.queue.get(block=True,timeout=int(delta))
                    test = 1
                except:
                    delta = delta * 2
                    if delta > 32:
                        test = 1
                        print("Failed")
                        return
                    else:
                        print("Retry ...\t")
            return response
```
and in the file coapclient.py 
```python
if op == "GET":
        if path is None:
            print "Path cannot be empty for a GET request"
            usage()
            sys.exit(2)
        response = client.get(path)
        if response is not None:
            print response.pretty_print()
        client.stop()
```
I'm not sure if it can help you but you can send some datagram if you haven't response with these changements.

**Comment by ashubhatt** [comment]
The issue still persists.. Any updates?

**Comment by ksrm** [comment]
Retransmissions and timeouts are a pretty fundamental part of CoAP, so I'm a bit surprised to see them not implemented. The get() method of the client class seems just to make a single call to protocol.send_message and wait indefinitely for a reply. 

Please don't take this the wrong way, but I think it's wrong to claim that this library is 'compliant with the RFC' until this is fixed - it's a bit misleading.

**Comment by Tanganelli** [comment]
Hi Kyle,
thanks for the feedback, but actually retransmissions are not handled
directly by client (that's why there is a single call on the client side),
instead the layered stack should take retransmissions into account in a
transparent way. It is a wired bug  because it has been raised by some of
the last changes I performed, in any case that I am trying to investigate
more deeply.

I'll let you know whenever I found a solution.

Best
Giacomo

2016-09-01 15:49 GMT+02:00 Kyle McInnes <notifications@github.com>:
